### PR TITLE
[AIRFLOW-1760] Password authentication for experimental rest API

### PR DESCRIPTION
**Pull request by NielsZeilemaker**
Modified the Password authentication to support HTTP Basic auth

Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1760

### Description
- [X] Here are some details about my PR, including screenshots of any UI changes:
Implemented HTTP Basic auth to allow Password authentication to be used to secure the experimental api.

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
None, the current passwor auth in contrib does not have any tests either

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by bolkedebruin** [comment]
Tests please and please quash your commits.

File: airflow/contrib/auth/backends/password_auth.py [hunk 1]
```diff
@@ -103,6 +104,21 @@ def load_user(userid):
     session.close()
     return PasswordUser(user)
 
+def is_valid(session, username, password):
```
Comment (bolkedebruin) [inline airflow/contrib/auth/backends/password_auth.py]: isn't there a better name? authenticate?

Comment (bolkedebruin) [reply airflow/contrib/auth/backends/password_auth.py]: And document your function please.

**Review by bolkedebruin — CHANGES_REQUESTED** [review]

**Comment by NielsZeilemaker** [comment]
@bolkedebruin I've added tests and squashed the commits

**Comment by NielsZeilemaker** [comment]
@bolkedebruin @Fokko would it be possible to merge this pull request?

**Comment by NielsZeilemaker** [comment]
@Fokko @bolkedebruin any comments? I would really like to get this merged

File: airflow/contrib/auth/backends/password_auth.py [hunk 2]
```diff
@@ -1,151 +1,212 @@
...
+
+from __future__ import unicode_literals
+
+from sys import version_info
+
+import base64
+import flask_login
+from flask_login import current_user
+from flask import flash, Response
+from wtforms import (
```
Comment (Fokko) [inline airflow/contrib/auth/backends/password_auth.py]: Please make this a single line and remove the brackets

File: airflow/contrib/auth/backends/password_auth.py [hunk 3]
```diff
@@ -1,151 +1,212 @@
...
+    Form, PasswordField, StringField)
+from wtforms.validators import InputRequired
+from functools import wraps
+
+from flask import url_for, redirect, make_response
+from flask_bcrypt import generate_password_hash, check_password_hash
+
+from sqlalchemy import (
```
Comment (Fokko) [inline airflow/contrib/auth/backends/password_auth.py]: Please make this a single line and remove the brackets

File: airflow/contrib/auth/backends/password_auth.py [hunk 4]
```diff
@@ -1,151 +1,212 @@
...
+    username = None
+    password = None
+
+    form = LoginForm(request.form)
+
+    if request.method == 'POST' and form.validate():
+        username = request.form.get("username")
+        password = request.form.get("password")
+
+    session = settings.Session()
```
Comment (Fokko) [inline airflow/contrib/auth/backends/password_auth.py]: You already have a session provided by the decorator.

**Review by Fokko — COMMENTED** [review]
Minor changes, please review, apart from that LGTM. Thanks for providing the tests.

**Comment by Fokko** [comment]
Thanks @NielsZeilemaker

**Pull request closed without merging**